Graphics/MaliCompilerReport/Assets/Editor/MaliShaderReport.cs

```diff
@@ -0,0 +1,644 @@
+                    passInfo.pass = subshader.GetPass(p);
+                    passInfo.name = passInfo.pass.Name;
+
+                    ParseKeywords(passInfo.pass.SourceCode, ref passInfo.keywords);
+
+                    // Create some label hints
+                    for (int i = 0; i < passInfo.keywords.Count; i++)
+                    {
+                        for (int j = 0; j < passInfo.keywords[i].Count; j++)
+                        {
+                            if (passInfo.keywords[i][j].Length > 0)
+                            {
+                                sb.Append(passInfo.keywords[i][j]);
+
+                                if (sb.Length > k_MaxHintLength)
+                                {
+                                    sb.Remove(k_MaxHintLength, sb.Length - k_MaxHintLength);
+                                    sb[15] = '.';
+                                    sb[16] = '.';
+                                    sb[17] = '.';
+
+                                    break;
+                                }
+                                else if (sb.Length == k_MaxHintLength)
+                                {
+                                    break;
+                                }
+                            }
+                        }
+
+                        passInfo.keywordHint.Add(sb.ToString());
+                        sb.Clear();
+                    }
+
+                    m_Passes.Add(passInfo);
+                    names.Add(passInfo.name);
+                }
+            }
+
+            if (m_SelectedPass < 0)
+                m_SelectedPass = 0;
+            else if (m_SelectedPass >= m_Passes.Count)
+            {
+                m_SelectedPass = m_Passes.Count - 1;
+            }
+
+            SetSelectedPass(m_SelectedPass, true);
+
+            m_PassNames = names.ToArray();
+        }
+    }
+
+    void SetSelectedPass(int index, bool init = false)
+    {
+        if (init || index != m_SelectedPass)
+        {
+            m_SelectedPass = index;
+
+            m_SelectedKeywords.Clear();
+
+            if (m_SelectedPass >= 0 && m_SelectedPass < m_Passes.Count)
+            {
+                PassInfo info = m_Passes[m_SelectedPass];
+
+                for (int i = 0; i < info.keywords.Count; i++)
+                {
+                    m_SelectedKeywords.Add(0);
+                }
+            }
+        }
+    }
+
+    void ParseKeywords(string code, ref List<List<string>> keywords)
+    {
+        keywords.Clear();
+
+        int pragmaIndex = code.IndexOf("#pragma");
+
+        // TODO: Change this to just use index of the string and avoid this substring nonscence as that just creating unnecessary garbage
+        while (pragmaIndex != -1)
+        {
+            int endOfLine = code.IndexOf('\n', pragmaIndex);
+            ParsePragmaString(code.Substring(pragmaIndex, endOfLine - pragmaIndex), ref keywords);
+            code = code.Substring(endOfLine + 1);
```

Comment:
You could also use `IndexOf(Char, Int32)` to start the string search after the `endOfLine` index instead of chopping the shader code smaller each time you move on to the next pragma.
https://docs.microsoft.com/en-us/dotnet/api/system.string.indexof?view=net-5.0#System_String_IndexOf_System_Char_System_Int32_